eclipse-collections/src/main/java/org/eclipse/collections/impl/map/mutable/UnifiedMap.java

```diff
@@ -1038,17 +1038,40 @@ public V removeKey(K key)
     @Override
     public boolean removeIf(Predicate2<? super K, ? super V> predicate)
     {
-        int previousSize = this.size();
-        Iterator<Entry<K, V>> iterator = this.entrySet().iterator();
-        while (iterator.hasNext())
+        int previousOccupied = this.occupied;
+        for (int index = 0; index < this.table.length; index += 2)
         {
-            Entry<K, V> entry = iterator.next();
-            if (predicate.accept(entry.getKey(), entry.getValue()))
+            Object cur = this.table[index];
+            if (cur == null)
+            {
+                continue;
+            }
+            if (cur == CHAINED_KEY)
             {
-                iterator.remove();
+                Object[] chain = (Object[]) this.table[index + 1];
+                for (int chIndex = 0; chIndex < chain.length; chIndex += 2)
+                {
+                    K key = this.nonSentinel(chain[chIndex]);
```

Comment:
there is a subtle bug here. It would be good to have a test for it. The bug is this: `chIndex` is incremented unconditionally, even if the predicate matches. So if we have 3 elements, `a`, `b`, and `c`, if `a` matches, the chain look likes `c`, `b`, but `c` is no longer put through the predicate because the index will move on.

Reply:
Ah, got it. Let me add a test case for it and fix it.

Reply:
Thanks for Pointing this out. Fixed it.

Reply:
Style nitpick... now with the refactoring of the for-loop, I think you could move the condition `chain[chIndex] != null` into the loop condition.